src/Hint/Lambda.hs

```diff
@@ -86,96 +93,179 @@ yes = map (\f -> dataDir </> f) dataFiles -- (dataDir </>)
 import Util
 import Data.List.Extra
-import Data.Maybe
 import qualified Data.Set as Set
 import Refact.Types hiding (RType(Match))
+import Data.Generics.Uniplate.Operations (universe, universeBi, transformBi)
 
+import BasicTypes
+import GHC.Util.Brackets (isAtom')
+import GHC.Util.FreeVars (free', allVars', freeVars', pvars', vars', varss')
+import GHC.Util.HsExpr (allowLeftSection, allowRightSection, niceLambdaR', lambda)
+import GHC.Util.Outputable
+import GHC.Util.RdrName (rdrNameStr')
+import GHC.Util.View
+import HsSyn
+import Language.Haskell.GhclibParserEx.GHC.Hs.Expr (isTypeApp, isOpApp, isLambda, isQuasiQuote, isVar, isDol)
+import OccName
+import RdrName
+import SrcLoc
 
-lambdaHint :: DeclHint
-lambdaHint _ _ x = concatMap (uncurry lambdaExp) (universeParentBi x) ++ concatMap lambdaDecl (universe x)
-
-
-lambdaDecl :: Decl_ -> [Idea]
-lambdaDecl (toFunBind -> o@(FunBind loc1 [Match _ name pats (UnGuardedRhs loc2 bod) bind]))
-    | isNothing bind, isLambda $ fromParen bod, null (universeBi pats :: [Exp_]) =
-      [warn "Redundant lambda" o (gen pats bod) [Replace Decl (toSS o) s1 t1]]
-    | length pats2 < length pats, pvars (drop (length pats2) pats) `disjoint` varss bind
-        = [warn "Eta reduce" (reform pats bod) (reform pats2 bod2)
-            [ -- Disabled, see apply-refact #3
-              -- Replace Decl (toSS $ reform pats bod) s2 t2]]
-            ]]
-        where reform p b = FunBind loc [Match an name p (UnGuardedRhs an b) Nothing]
-              loc = setSpanInfoEnd loc1 $ srcSpanEnd $ srcInfoSpan loc2
-              gen ps = uncurry reform . fromLambda . Lambda an ps
-              (finalpats, body) = fromLambda . Lambda an pats $ bod
-              (pats2, bod2) = etaReduce pats bod
-              template fps b = prettyPrint $ reform (zipWith munge ['a'..'z'] fps) (toNamed "body")
-              munge :: Char -> Pat_ -> Pat_
-              munge ident p@(PWildCard _) = p
-              munge ident p = PVar (ann p) (Ident (ann p) [ident])
-              subts fps b = ("body", toSS b) : zipWith (\x y -> ([x],y)) ['a'..'z'] (map toSS fps)
-              s1 = subts finalpats body
-              --s2 = subts pats2 bod2
-              t1 = template finalpats body
-              --t2 = template pats2 bod2
-
-lambdaDecl _ = []
-
-setSpanInfoEnd ssi (line, col) = ssi{srcInfoSpan = (srcInfoSpan ssi){srcSpanEndLine=line, srcSpanEndColumn=col}}
-
-
-etaReduce :: [Pat_] -> Exp_ -> ([Pat_], Exp_)
-etaReduce ps (App _ x (Var _ (UnQual _ (Ident _ y))))
-    | ps /= [], PVar _ (Ident _ p) <- last ps, p == y, p /= "mr", y `notElem` vars x
-    , not $ any isQuasiQuote $ universe x
-    = etaReduce (init ps) x
-etaReduce ps (InfixApp a x (isDol -> True) y) = etaReduce ps (App a x y)
-etaReduce ps x = (ps,x)
+lambdaHint :: DeclHint'
+lambdaHint _ _ x
+    =  concatMap (uncurry lambdaExp') (universeParentBi x)
+    ++ concatMap lambdaDecl' (universe x)
+
+-- TODO: handle PatBinds
```

Comment:
I think this TODO can safely be deleted, as per the comments in the pull request

Reply:
Then please delete :)

Reply:
Done